Adds support for all Consul 0.8 ACL rule scopes

colin-nolan opened the pull request:
##### SUMMARY
Adds support for the `consul_acl` module to set all ACL rule scopes available in Consul 0.8:
https://www.consul.io/docs/guides/acl.html#acl-rules-and-scope

Builds on the work by @nathanwebsterdotme in #23467.

Fixes: #19341.


##### ISSUE TYPE
 - Feature Pull Request


##### COMPONENT NAME
consul_acl.py

##### ANSIBLE VERSION
```
ansible 2.4.0
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.5/dist-packages/ansible-2.4.0-py3.5.egg/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.5.2 (default, Nov 17 2016, 17:05:23) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
Implementation has required the refactoring of a large part of the `consul_acl`. Whilst at it, I have also cleaned up and made various improvements.

Commit:
Added in support for 'agent' and 'node' types.

Commit:
Tidies and moves `consul_acl` module closer to PEP8 compliance.

Commit:
Switched from using byspoke code to handle py2/3 string issues to using `to_text`.

Commit:
Made changes suggested by jrandall in https://github.com/ansible/ansible/pull/23467#pullrequestreview-34021967.

Commit:
Refactored consul_acl to support scopes with no pattern (and therefore a different HCL defintion).

Commit:
Corrects whitespace in Consul ACL HCL representation.

Commit:
Fixes Consul ACL to return the HCL equivalent JSON (according to the Consul docs) for the set ACLs.

colin-nolan commented:
AFAIK this is now stable and will work. I'll finish up the loose ends and test it out a bit more (e.g. with Py2) next week.

I assume the most appropriate testing would involve adding some more ACL focused tasks here:
https://github.com/wtsi-hgi/ansible/blob/a8d0e64080e7fce9ccd568b0cd86eb83cad6f50a/test/integration/consul.yml

Commit:
Repositioned import to align with Ansible standard (!= PEP8 standard).

Commit:
Adds Python 2.6 compatibility.

Commit:
Fixes PEP8 issues.

Commit:
Removes consul_acl.py as it now passes PEP8.

Commit:
Follows advice in the "Documenting Your Module" guide and moves imports up from the bottom.

Commit:
Tidies consul_acl module documentation.

Commit:
Updates link to guide about Consul ACLs.

Commit:
Removes new line spaces from error message string.

Commit:
Provide better error message if user forgets to associate a value to a Consul ACL rule.

Commit:
Minor refactoring of Consul ACL module.

Commit:
Fixes bug that was breaking idempotence in Consul ACL module.

Commit:
Detects redefinition of same rule.

Commit:
Adds test to check the Consul ACL module can set rules for all supported scopes.

sgargan commented on lib/ansible/modules/clustering/consul_acl.py:
```diff
@@ -151,56 +158,48 @@ def execute(module):
         module.fail_json(msg="Could not create/update acl %s" % e)
 
+    # TODO: It is unclear to me why `token`, `rules`, `name` and `type` are returned - they are all inputs. The only one
+    # I can see being useful is the JSON view of the rules.
```
The token is supplied when the acl is created so should be returned. As you mention the rules are probably more useful returned as json. The others are inputs and so it's probably not necessary to return them.

colin-nolan replied:
Good point - I overlooked that the token can be created. I suggest that we take out `name` and `token_type`, given that they're not likely to be useful in the return.

I would also like to float the idea that we return the remaining rules when an ACL is removed. Although this would require an extra API call, I can see it being helpful (particularly when debugging) and it would make the return consistent.

I'll complete the `RETURN` documentation once we come to a conclusion.

sgargan reviewed (commented):
Changes look good, thanks for taking the time to add them. There are existing integration tests for the consul module in [core](https://github.com/ansible/ansible/blob/devel/test/integration/roles/test_consul_acl/tasks/main.yml).  Given how much has changed here it would be extremely useful to get the examples you've specified added as tests.

colin-nolan commented:
@sgargan many thanks for your comments. I did add what could very loosely be described as a test to the [`consul.yml` file I found in the integration tests directory](https://github.com/wtsi-hgi/ansible/blob/3b09cd5bda82383bc5a270c157731b35e5566d20/test/integration/consul.yml#L70-L96)  - I'll move that over to the module you linked to instead, which looks much more appropriate.

Commit:
Fixes return when updating an ACL.

Commit:
Clean up of Consul ACL integration test file.

Commit:
Verify correct changes to existing Consul ACL rule.

Commit:
Adds tests for idempotence.

Commit:
Splits Consul ACL tests into cohesive modules.

Commit:
Adds test for deleting Consul ACLs.

Commit:
Test that Consul ACL module can set all rule scopes.

manos commented on lib/ansible/modules/clustering/consul_acl.py:
```diff
@@ -151,56 +158,48 @@ def execute(module):
             existing_rules = load_rules_for_token(module, consul, token)
-            supplied_rules = yml_to_rules(module, rules)
-            changed = not existing_rules == supplied_rules
+            changed = existing_rules != rules
```
Hi! Thanks for this... exactly what I needed.
I'm unsure how this is passing tests, as ansible detects a change each time I run. Looks like you're comparing HCL and JSON here.
If I change this line to:
            changed = encode_rules_as_json(existing_rules) != encode_rules_as_json(rules)
it works as expected (changes when I change a rule, otherwise it doesn't).

colin-nolan replied:
The reason it passed [the tests](https://github.com/ansible/ansible/blob/devel/test/integration/roles/test_consul_acl/tasks/main.yml) was because they were rather minimial and did not test for idempotence ;).

I spent some time yesterday [improving them](https://github.com/wtsi-hgi/ansible/tree/feature/consul_acl_rules/test/integration/roles/test_consul_acl/tasks) and the issue you describe *should* now cause them to fail.

However, for some reason the [CI testing seems to be passing just fine](https://app.shippable.com/github/ansible/ansible/runs/26238/summary/console)? If I run the `test_consul_acl` independently on the Ansible test setup, they fail as expected though. If anyone can spot anything wrong with the way these tests are setup, please let me know.

colin-nolan replied:
>  Looks like you're comparing HCL and JSON here.

I don't think that's true: it (should) be comparing two `RuleCollection` instances.


> ...as ansible detects a change each time I run. 

The tests reveal that this happens when setting a new token - there is no check for whether a token with the same name has been created before. I'll fix this, along with any other idempotence issues that lie after this one. After I fix the issues highlighted in the new tests, it would be great if you could verify the fixes with what you're running. I'll let you know when I'm done.

manos replied:
@colin-nolan sure, I'm happy to help test.

manos commented:
Two other notes: 
1) this doesn't seem to work with operator or keyring rules. These are different, in that they don't specify: type "path" { policy = "write"}, they simply want: "operator": "write" in the hcl.
If these rules exist in the ACL already, consul_acl fails to load existing rules. If you try to specify them in the playbook, it also fails due to formatting differences.

2) if the rule doesn't exist, but the user provided a token, this fails to create a new ACL because load_rules_for_token() gets None back from the API, but doesn't check for that. Here is my diff that works for me:
```
@@ -172,7 +173,11 @@
     try:
         if token:
             existing_rules = load_rules_for_token(module, consul, token)
-            changed = existing_rules != rules
+        else:
+            existing_rules = {}
+        if token:
+            changed = encode_rules_as_json(existing_rules) != encode_rules_as_json(rules)
+            #changed = existing_rules != rules
             if changed:
                 token = consul.acl.update(
                     token,
@@ -214,6 +219,8 @@
 def load_rules_for_token(module, consul_api, token):
     try:
         info = consul_api.acl.info(token)
+        if not info:
+            return {}
         rules_as_hcl_string = to_text(info['Rules'])
         rules_as_json = hcl.loads(rules_as_hcl_string)
         return decode_rules_as_json(rules_as_json)
```

Commit:
Fixes issues surrounding the creation of ACLs.

Thanks for the comments by manos in https://github.com/ansible/ansible/pull/25800#issuecomment-310137889.

Commit:
Stops Consul ACL's name being "forgotten" if ACL updated by token.

Commit:
Fixes incorrect assignment when a Consul ACL is deleted.

Commit:
Fixes value of `changed` when Consul ACL is removed.

Commit:
Fixes tests for Consul ACL.

Commit:
Adds interal documentation.

Commit:
Refactors to separate update and create (also makes it possible to unit test this module).

Commit:
Improves documentation.

Commit:
Completes RETURN documentation for Consul ACL module.

colin-nolan commented:
I am happy with this now.

ready_for_review

manos commented:
@colin-nolan either I'm testing the wrong version, or idempotentcy, isn't.
I'm using this version of the file: https://raw.githubusercontent.com/wtsi-hgi/ansible/599e30ada1ee97f099acdfbb7e151cedcf2a6e5a/lib/ansible/modules/clustering/consul_acl.py

It seems to update ACLs on every run. I target three (server) machines, and normally (with my hacked version) if an ACL changes, the first server will update it, and the other two won't. Running the version linked above, all three servers detect a change and re-write the ACLs on every run.

colin-nolan commented:
@manos so you're updating the same ACL on 3 servers in parallel/in quick succession? If so then I think the behavour that you described is to be expected...

If the same request to create a new token gets send to all servers in parallel, they will all check for its existence (for which they'll find it doesn't exist) and then start the process of updating the value. Therefore, all related Ansible tasks will report a change.

The only way to make this module safe to run concurrently like that would be to use locks that are shared across the cluster. Unless you have implemented this in your hacked version, I suspect that it works because it doesn't update each server in parallel.

Given that ACLs will get synchronised across the cluster, I don't think it makes sense to set the same ACLs on all servers. For example, we just delegate the setting of post-bootstrapping ACLs to the Consul leader. The only time you need to set on each server is if it is not yet connected to the cluster and you're setting the ACLs required to get them connected, in which case it makes sense that the task would report a change on each server.

Does the above explain the issue? Or are you are finding that setting ACLs sequentially is not idempotent.

manos commented:
Sorry. Please ignore that detail. If I constrain my run to a single node, yes, it updates the same ACLs on every single run :)

colin-nolan commented:
@manos can you post an example showing this problem? I have added a test for idempotence but it works fine when I test it: https://github.com/wtsi-hgi/ansible/blob/599e30ada1ee97f099acdfbb7e151cedcf2a6e5a/test/integration/roles/test_consul_acl/tasks/update-acl.yml#L33-L48. Many thanks for your time.

manos commented:
@colin-nolan I can't currently make sense of why it's detecting changes every time. I added an exception to print out the existing and new rules:

changed = existing_acl.rules != configuration.rules
raise Exception({"existing": str(existing_acl.rules), "new": str(configuration.rules), "changed": changed})

yields:
{'new': 'node "" {\n  policy = "write"\n}\nservice "" {\n  policy = "write"\n}\nagent "" {\n  policy = "write"\n}\nsession "" {\n  policy = "write"\n}\nkey "" {\n  policy = "write"\n}\nquery "" {\n  policy = "write"\n}\nevent "" {\n  policy = "write"\n}\n', 
'changed': True, 
'existing': 'node "" {\n  policy = "write"\n}\nservice "" {\n  policy = "write"\n}\nagent "" {\n  policy = "write"\n}\nsession "" {\n  policy = "write"\n}\nkey "" {\n  policy = "write"\n}\nquery "" {\n  policy = "write"\n}\nevent "" {\n  policy = "write"\n}\n'
}

This particular rules looks like:
    - name : infrastructure acl
      consul_acl:
        mgmt_token: "{{ consul_acl_master_token }}"
        token: "{{ consul_infrastructure_client_token }}"
        token_type: client
        name: "full RW access to consul"
        rules:
          - key: ""
            policy: write
          - service: ""
            policy: write
          - event: ""
            policy: write
          - query: ""
            policy: write
          - node: ""
            policy: write
          - session: ""
            policy: write
          - agent: ""
            policy: write
        state: present

But simpler rules all trigger this too.
They look the same, but changed is always being set to True on line # 228.

Commit:
Fixes issue with equality checking for `None` in ACL Consul.

Commit:
Fixes Python 2 issue with making a decision based on `str` type.

Commit:
Fixes inequality check bug in Python 2.

manos commented:
Boom!
@colin-nolan that did it. All is working for me now.
Thanks much. This module is super useful... don't have to write my own provisioning scripts, as I do for vault.

colin-nolan commented:
Great! :D

The main issue only affected Python 2, which I usually avoid and hence why I wasn't seeing the problem. It turns out that in the 8 year outdated version of the language, the following doesn't do what you think it might (it requires the `__ne__` method to be overriden instead):
```
class A():
  def __eq__(self, other):
    print("here")
    
A() != A()
```

On a related note, I have no confidence that the tests for this module are actually being ran correctly in the CI - the CI always comes back green, even when it definitely should have failed! If it was running with both Py2 and Py3, this issue would have been flagged up by the tests that I put in. However, unless there is something obviously wrong with the way I've added the new tests, this issue is beyond the scope of this PR.

sgargan commented:
It looks good @colin-nolan, I'm curious that your tests are not running in ci though. How are you running them locally?

sgargan commented:
its good to :ship: shipit

colin-nolan commented:
@sgargan I've added a bug report for the CI issue: https://github.com/ansible/ansible/issues/26229.

After losing faith in the CI, I created a crude playbook to run the test role against a test cluster during development.

manos commented:
@colin-nolan So, I tried using this new version to push my ACLs to prod last night. There were no ACLs written to consul at all.
It blew up.. errors about "existing token does not exist." (paraphrased)
Sorry, it's long gone from my scrollback, as I was trying to recover the infrastructure (I had default deny policy - so that took effect with 0 ACLs being written).

I suspect starting clean (with no ACLs), this is easily reproducible by specifying a token for a new rule? I will try to test that later today in my test cluster.

colin-nolan commented:
@manos if this wasn't obviously a problem with the implementation, it would be great if the merge goes ahead. I'm confident that this PR is an improvement on the current implementation, both in terms of functionality and correctness. There may still be some hidden bugs (what software doesn't have them ;)) but it's likely they also exist in the original implementation. I feel it would be best to switch over to this work and then iterate on it from there.

@sgargan what is the process to get this merged? I would very much like to move forwards with this as it's been a lot of work.

jctanner commented:
bot_status

ascendantlogic commented:
So I'm using this in my own project to manage ACL's since I need `operator` and `node` rules. Previously I had generated my own UUIDs and put them in a vault-encrypted vars file and passed them in to the older module to control them and be able to reference them for usage after the ACLs were created. Attempting to do that now results in the `ConsulACLNotFoundException` being thrown because my use of a token value makes the module think I'm updating an existing ACL. However my attempts to update the `create_acl` function to pass in a pre-existing token value is also throwing errors. Is there a way to either create ACLs with predefined token values or to dump the value of the created token to the ansible output? I feel like I'm missing something here.

colin-nolan commented:
@ascendantlogic
> Previously I had generated my own UUIDs and put them in a vault-encrypted vars file and passed them in to the older module to control them and be able to reference them for usage after the ACLs were created. Attempting to do that now results in the ConsulACLNotFoundException being thrown because my use of a token value makes the module think I'm updating an existing ACL. 

Am I correct to assume that at this point you haven't yet created the ACL associated to that token? Is so, that exception is expected as, if a token is supplied, it assumes the operation is an update. This behaviour is the same as that in [the previous implementation](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/clustering/consul_acl.py#L166-L186) (I can't see how this would have worked for you previously as you can see that, if there's a token, the update branch is taken, opposed to create?).

> However my attempts to update the create_acl function to pass in a pre-existing token value is also throwing errors. Is there a way to either create ACLs with predefined token values...

It looks like it's possible to create an ACL with a given token (acl_id) (http://python-consul.readthedocs.io/en/latest/#consul.base.Consul.ACL.create) so it should be achievable.

>  ... or to dump the value of the created token to the ansible output?

The token associated with the ACL is returned: https://github.com/wtsi-hgi/ansible/blob/7f2b3003ff570515f62ec0496e2dcc9b42b41bf8/lib/ansible/modules/clustering/consul_acl.py#L137-L158.

manos commented:
With tokens that don't yet exist, I don't understand how the previous implementation is working either, but it is (that's exactly the problem I ran into).

acasademont commented:
@colin-nolan I think you mentioned the wrong guy :p @ascendantlogic was probably the intended one!

colin-nolan commented:
Sorry @acasademont - clearly I wasn't paying enough attention to the autocomplete!

colin-nolan commented:
@manos - thanks for the confirming this. I can only assume that this works in the original implemenetation by slipping down the update path and suprisingly not tripping up as a result. Looking further, this example does implies that this functionality is supported: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/clustering/consul_acl.py#L105-L116

@manos, @ascendantlogic: I can add support for setting the ACL with a given token. I'm not sure when I'll time for this yet though (contributions welcome!). I'm also rather concerned that there's a lot of work in this one PR. I'd much prefer it if this PR is accepted as is (with existing maintainer approval) and then this extra functionality is added in a subsequent PR.

Commit:
Adds tests for setting ACL with token.

Commit:
Adds support for creating an ACL with a given token.

Commit:
Outputs operation performed on Consul ACL when changed.

Commit:
Fixs issue with test for creating a Consul ACL with rules.

Commit:
Corrects property used to set ACL token in python-consul library.

Commit:
Fixes tear-down issue in test that creates a Consul ACL using a token.

colin-nolan commented:
@manos, @ascendantlogic: I've added support for setting the token upon creation - it would be great if you could see if it does what you expect. I've not got time for getting the PR ready for merge at the moment (there's now conflicts).

manos commented:
@colin-nolan just tested by deleting a token in my test env, and watched ansible re-create it. Yay \o/ this is awesome.

colin-nolan commented:
@sgargan please may you re-approve?

I would very much like to get this merged ASAP as I can't keep maintaining this PR forever...

sgargan commented:
Nice one Colin! Thanks for all your efforts with this. shipit

samdoran commented:
shipit
bot_status
